Can't Validate Keys with Dots in Them

It seems that I can't validate a YAML file if it has dots (`.`) in the key.  Looks like it's because of the flattening of keys later on, but interestingly you _can_ specify a key with dots in the schema, and Yamale will translate the schema keys with dots in them to underscores.

I worked around this in a _super_ hacky way locally with the following:

```python
from yamale import util

def sub(dic):
    new = {}
    for k, v in util.get_iter(dic):
        if util.is_iter(v):
            v = sub(v)
        if util.isstr(k):
            new[k.replace('.', '_')] = v
    return new


def parse_file(file_name, parser):
    try:
        parse = _parsers[parser.lower()]
    except KeyError:
        raise NameError('Parser "' + parser + '" is not supported\nAvailable parsers are listed below:\nPyYAML\nruamel')
    parsed = parse(file_name)
    new = []
    for x in parsed:
        new.append(sub(x))
    return new
```

This pretty much just replaces `.` with `_` in parsed YAML files, which is being done elsewhere anyway when things get flattened out.  I'm not sure if the flattening functionality that does the dot replacement could be removed with this patch, but maybe it could be.

Would you be open to a cleaned up and tested version of this in a PR?  If so, I can work on it and submit it.

My primary focus is networking, so this is useful when I need to validate a document with IP addresses or MAC addresses as a key.

While writing new functional test cases, I discovered that this issue is actually more complicated than I thought.  Everything seems fine when the value is a scalar, but when it's a mapping, there's a `KeyError` that's raised.  Given this schema:

```yaml
---
ntp:
  config:
    enabled: bool()
    ntp-source-address: str()
  servers:
    "192.168.1.1":
      str: str()
    "192.168.1.2":
      str: str()
```

And this input:

```yaml
---
ntp_server_profile_generic: &ntp_profile_generic
  port: 123
  version: 3
  prefer: true
ntp:
  config:
    enabled: true
    ntp-source-address: 1.2.3.4
  servers:
    192.168.1.1:
      str: yes
    192.168.1.2:
      str: no
```

The following happens when running tests:

```
============================================================== ERRORS ===============================================================
_________________________________________ ERROR collecting yamale/tests/test_functional.py __________________________________________
yamale/tests/test_functional.py:80: in <module>
    d[key] = yamale.make_schema(get_fixture(d[key]))
yamale/yamale.py:17: in make_schema
    s = Schema(raw_schemas[0], path, validators=validators)
yamale/schema/schema.py:18: in __init__
    self._schema = self._process_schema(schema_dict, self.validators)
yamale/schema/schema.py:34: in _process_schema
    schema_flat = util.flatten(schema_dict)
yamale/util.py:41: in flatten
    child.update(flatten(dic[k], keep_iter, item_position))
yamale/util.py:41: in flatten
    child.update(flatten(dic[k], keep_iter, item_position))
yamale/util.py:41: in flatten
    child.update(flatten(dic[k], keep_iter, item_position))
E   KeyError: '192_168_1_1'
```

However, the following is successful:

Schema:

```yaml
---
ntp:
  config:
    enabled: bool()
    ntp-source-address: str()
  servers:
    "192.168.1.1": str()
    "192.168.1.2": str()
```

Input:

```yaml
---
ntp_server_profile_generic: &ntp_profile_generic
  port: 123
  version: 3
  prefer: true
ntp:
  config:
    enabled: true
    ntp-source-address: 1.2.3.4
  servers:
    192.168.1.1: abcd
    192.168.1.2: efgh
```

So the issue isn't quite where I thought it was, though my changes fix it anyway.  I'm going to submit the patch I've written as a PR, and we can talk about different ways to fix the issue in that PR.

--- a/yamale/schema/schema.py
+++ b/yamale/schema/schema.py
@@ -67,7 +67,7 @@ class Schema(object):
         errors = []
 
         if position:
-            position = '%s.%s' % (position, key)
+            position = '%s%s%s' % (position, util.YAMALE_SEP, key)
         else:
             position = key
 

--- a/yamale/tests/fixtures/issue_50.yaml
+++ b/yamale/tests/fixtures/issue_50.yaml
@@ -1,0 +1,16 @@
+---
+ntp:
+  config:
+    enabled: bool()
+    ntp-source-address: str()
+  servers:
+    "192.168.1.1": include('ntp_servers')
+    "192.168.1.2": include('ntp_servers')
+---
+ntp_servers:
+  config: include('ntp_servers_config')
+ntp_servers_config:
+  port: int(min=123, max=123)
+  version: int(min=2, max=3)
+  prefer: bool()
+  address: str()

--- a/yamale/tests/fixtures/issue_50_good.yaml
+++ b/yamale/tests/fixtures/issue_50_good.yaml
@@ -1,0 +1,18 @@
+---
+ntp_server_profile_generic: &ntp_profile_generic
+  port: 123
+  version: 3
+  prefer: true
+ntp:
+  config:
+    enabled: true
+    ntp-source-address: 1.2.3.4
+  servers:
+    192.168.1.1:
+      config:
+        <<: *ntp_profile_generic
+        address: 192.168.1.1
+    192.168.1.2:
+      config:
+        <<: *ntp_profile_generic
+        address: 192.168.1.2

--- a/yamale/tests/test_functional.py
+++ b/yamale/tests/test_functional.py
@@ -1,7 +1,10 @@
 import pytest
 import yamale
+
 from . import get_fixture
 from .. import validators as val
+from yamale.util import YAMALE_SEP
+
 
 types = {
     'schema': 'types.yaml',
@@ -56,6 +59,11 @@ issue_22 = {
     'good': 'issue_22_good.yaml'
 }
 
+issue_50 = {
+    'schema': 'issue_50.yaml',
+    'good': 'issue_50_good.yaml'
+}
+
 regexes = {
     'schema': 'regex.yaml',
     'bad': 'regex_bad.yaml',
@@ -78,7 +86,7 @@ test_data = [
     types, nested, custom,
     keywords, lists, maps,
     anys, list_include, issue_22,
-    regexes, ips, macs
+    issue_50, regexes, ips, macs
 ]
 
 for d in test_data:
@@ -102,7 +110,7 @@ def test_nested_schema():
     nested_schema = nested['schema']
     assert isinstance(nested_schema['string'], val.String)
     assert isinstance(nested_schema.dict['list'], (list, tuple))
-    assert isinstance(nested_schema['list.0'], val.String)
+    assert isinstance(nested_schema['list%s0' % YAMALE_SEP], val.String)
 
 
 @pytest.mark.parametrize('data_map', test_data)

--- a/yamale/util.py
+++ b/yamale/util.py
@@ -9,6 +9,9 @@ except ImportError:
     from collections import Mapping, Set, Sequence
 
 
+YAMALE_SEP = '`/4M4|_E'
+
+
 # Python 3 has no basestring, lets test it.
 try:
     basestring  # attempt to evaluate basestring
@@ -37,10 +40,8 @@ def flatten(dic, keep_iter=False, position=None):
         return {}
 
     for k, v in get_iter(dic):
-        if isstr(k):
-            k = k.replace('.', '_')
         if position:
-            item_position = '%s.%s' % (position, k)
+            item_position = '%s%s%s' % (position, YAMALE_SEP, k)
         else:
             item_position = '%s' % k
 
@@ -58,7 +59,7 @@ def get_expanded_path(dic, key):
     path = []
     cur = dic
     try:
-        keys = key.split('.')
+        keys = key.split(YAMALE_SEP)
     except AttributeError:
         keys = [key]
     for k in keys[:-1]:

